### PR TITLE
fix aoflux agrid/xgrid problems found in testing

### PR DESCRIPTION
**Pull request by mvertens**
### Description of changes
bug fixes for new agrid/ogrid and restart functionality

### Specific notes

1. add fixes for computing aofluxes on both agrid and xgrid in compsets with data ocean and active atm (i.e. F/Q compsets in CESM)
2. fix problem associated with running ERS/ERR tests with new introduction of write_restart_at_endofrun

Contributors other than yourself, if any:

CMEPS Issues Fixed: None (no issues were raised for the problems being fixed)

Are changes expected to change answers?
 - [x] bit for bit
 - [ ] different at roundoff level
 - [ ] more substantial

Any User Interface Changes (namelist or namelist defaults changes)?
 - [x] Yes - write_restart_at_endofrun default has been changed
 - [ ] No

Testing performed if application target is CESM:(either UFS-S2S or CESM testing is required):
- [ ] (recommended) CIME_DRIVER=nuopc scripts_regression_tests.py
   - machines:
   - details (e.g. failed tests):
- [x] (recommended) CESM testlist_drv.xml
   - machines and compilers: cheyenne/intel
   - details (e.g. failed tests): 
     -  bfb compared to baseline  cmeps0.13.22, 
     - verifed that SMS_Ld7_Vnuopc.f09_g17.B1850.cheyenne_intel.allactive-defaultio cesm2_3_alpha05b_nuopc
- [ ] (optional) CESM prealpha test
   - machines and compilers
   - details (e.g. failed tests):
- [ ] (other) please described in detail
   - machines and compilers
   - details (e.g. failed tests):

Testing performed if application target is UFS-coupled:
- [ ] (recommended) UFS-coupled testing
   - description:
   - details (e.g. failed tests):

Testing performed if application target is UFS-HAFS:
- [ ] (recommended) UFS-HAFS testing
   - description:
   - details (e.g. failed tests):

Hashes used for testing:
- [x] CESM:
  - repository to check out: https://github.com/ESCOMP/CESM.git
  - hashes: cesm2_3_alpha05b with the following changes:
       cime: cime6.0.5  
       cism: cismwrap_2_1_87
       cdeps: cdeps0.12.20
  - [ ] UFS-coupled, then umbrella repostiory to check out and associated hash:
  - repository to check out:
  - branch:
  - hash:
- [ ] UFS-HAFS, then umbrella repostiory to check out and associated hash:
  - repository to check out:
  - branch:
  - hash:

### CODE REVIEW

**Comment by mvertens** [comment]
@denise @uturuncoglu  - please do not test this PR until I have had a chance to run the cesm cmeps tests and resolve all the issues.

**Comment by mvertens** [comment]
@uturuncoglu @DeniseWorthen - this PR is ready for review. For some reason github will not let me change the status - it gives me an error. Can you both please test this and make sure the changes do not change your answers. I'm happy to resolve any issues you run into. I'm also happy to have you both comment on the PR if you have questions.

**Comment by mvertens** [comment]
Actually, github let me change the status now for ready for review. Please give it a look.

**Comment by uturuncoglu** [comment]
@mvertens I run the HAFS coupled case but I am getting following error,

```
20210811 104641.993 INFO             PET323   med_fldList_GetFldNames: ERROR either flds or fldnames have not been allocate
20210811 104641.993 ERROR            PET323 med_phases_aofluxes_mod.F90:169   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 med.F90:2151   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 MED:src/addon/NUOPC/src/NUOPC_ModelBase.F90:1842   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 EARTH Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:2587   Failure  - Phase 'IPDv03p7' Initialize for modelComp 1: MED did not return ESMF_SUCCESS
20210811 104641.993 ERROR            PET323 EARTH Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:2418   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 EARTH Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:2202   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 EARTH Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:473   Failure  - Passing error in return code
20210811 104641.993 ERROR            PET323 MAIN_NEMS.F90:429   Failure  - Aborting NEMS
20210811 104641.993 INFO             PET323   Finalizing ESMF
```

Do I need. to add any extra option to the nems configure to make it run? maybe I need to add the option related with xgrid. Right?

**Comment by mvertens** [comment]
@uturuncoglu - thanks for testing this. I know what is wrong. Let me push some fixes back and you can test again.

**Comment by mvertens** [comment]
@uturuncoglu - I have provided a fix that hopefully will address your problem. Can you please test again and let me know if that works.

**Comment by uturuncoglu** [comment]
@mvertens thanks for the fix. I'll update you after testing again.

**Comment by uturuncoglu** [comment]
@mvertens it seems that your fix is working and I could run the test. BTW, I also need to add following to the nems.configure but this is related with recent @jedwards4b PR for the CDEPS.

```
  stop_n = 6
  stop_option = nhours
  stop_ymd = -999
```

Let me do couple of more test and I'll update you. BTW, I need to go to NCAR to replace my laptop this afternoon bu I'll test it after I return to home.

File: mediator/med.F90 [hunk 1]
```diff
@@ -2128,14 +2120,39 @@ subroutine DataInitialize(gcomp, rc)
...
+      !---------------------------------------
 
+      ! NOTE: this section must be done BEFORE the second call to esmFldsExchange
+      ! Create field bundles for mediator ocean albedo computation
+
+      fieldCount = med_fldList_GetNumFlds(fldListMed_aoflux)
+      if ( fieldCount > 0 ) then
+         if ( is_local%wrap%med_coupling_active(compocn,compatm) .or. &
+              is_local%wrap%med_coupling_active(compatm,compocn)) then
+            is_local%wrap%do_med_aoflux = .true.
```
Comment (uturuncoglu) [inline mediator/med.F90:2137]: Is this logic correct? `is_local%wrap%do_med_aoflux` is set as `true` in here but in line 2148 it sets to `false`. There is nothing between them related with setting `do_med_aoflux`.

**Review by uturuncoglu — COMMENTED** [review]
@mvertens it looks good to me. I have just one comment in my review. I'll also perform one more test and let you know. Thanks.

**Comment by DeniseWorthen** [comment]
I had to add the same changes for the coupled and datm tests for ufs-weather that Ufuk pointed out (related to stop_n etc) but otherwise all the test pass in ufs-weather.

**Review by DeniseWorthen — APPROVED** [review]

File: mediator/med.F90 (hunk 1 above)
Comment (mvertens) [reply mediator/med.F90]: @uturuncoglu - you are right. It turns out that is_local%wrap%do_med_aoflux is not used any more and so I have now removed it. I have just pushed my changes back with this removed. The rest of the logic should be correct.

**Comment by uturuncoglu** [comment]
@mvertens The uni-directional ATM-WAV also works. So, it looks fine and merged to master.

**Review by uturuncoglu — APPROVED** [review]

**Comment by mvertens** [comment]
@uturuncoglu @DeniseWorthen 
Thanks so much for your help with this PR!

**Pull request merged**